False positive on convenience_type for classes with static @objc methods

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

The `convenience_type` rule triggers when you have a class that has only class or static members, but it should necessarily:

1. You might be using `class` functions intentionally so that you can override them in a subclass.
2. If the functions are tagged with `@objc`, which they might be in order to use them as actions for controls or gesture recognizers, you can't change the class to an enum, because enums can't have `@objc` members.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ Pods/SwiftLint/swiftlint lint
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths 
Linting 'DrawingTests.swift' (1/125)
Linting 'MBTATests.swift' (2/125)
Linting 'TestConstants.swift' (3/125)
Linting 'MBTAViewTests.swift' (4/125)
Linting 'WeatherTests.swift' (5/125)
Linting 'Payloads.swift' (6/125)
Linting 'ArrayOfPointsTests.swift' (7/125)
Linting 'RectTests.swift' (8/125)
Linting 'CollectionTests.swift' (9/125)
Linting 'TestClient.swift' (10/125)
Linting 'FloatingPointTests.swift' (11/125)
Linting 'AppCoordinator.swift' (12/125)
Linting 'ContentCoordinator.swift' (13/125)
Linting 'WeatherCoordinator.swift' (14/125)
Linting 'MBTACoordinator.swift' (15/125)
Linting 'Coordinator.swift' (16/125)
Linting 'Fonts.swift' (17/125)
Linting 'ArrayOfPoints+Distance.swift' (18/125)
Linting 'UIColor+Extensions.swift' (19/125)
Linting 'ContentViewController.swift' (20/125)
Linting 'MBTAViewController.swift' (21/125)
Linting 'MBTAHeaderView.swift' (22/125)
Linting 'BusHeaderView.swift' (23/125)
Linting 'MBTARouteView.swift' (24/125)
Linting 'LogViewController.swift' (25/125)
Linting 'WeatherViewModel.swift' (26/125)
Linting 'WeatherCell.swift' (27/125)
Linting 'ForecastBackgroundView.swift' (28/125)
Linting 'ForecastCell.swift' (29/125)
Linting 'TemperatureCell.swift' (30/125)
Linting 'SunIntervalView.swift' (31/125)
Linting 'AspectImageCell.swift' (32/125)
Linting 'ForecastBackgroundViewModel.swift' (33/125)
Linting 'WeatherViewController.swift' (34/125)
Linting 'ClockView.swift' (35/125)
Linting 'GudTayView.swift' (36/125)
Linting 'CGPath+Utilities.swift' (37/125)
Linting 'DoodleView.swift' (38/125)
Linting 'DoodleService.swift' (39/125)
Linting 'DoodleViewModel.swift' (40/125)
Linting 'ImageIO.swift' (41/125)
Linting 'DoodleViewController.swift' (42/125)
Linting 'ToggleButton.swift' (43/125)
Linting 'Brush.swift' (44/125)
Linting 'CATransaction+Utilities.swift' (45/125)
Linting 'UIView+Extensions.swift' (46/125)
Linting 'UserDefaults+Utility.swift' (47/125)
Linting 'CGContext+Utilities.swift' (48/125)
Linting 'Array+Utilities.swift' (49/125)
Linting 'ClockTime.swift' (50/125)
Linting 'Date+Utilities.swift' (51/125)
Linting 'UICollectionReusableView+Utilities.swift' (52/125)
Linting 'ModalDismissBehavior.swift' (53/125)
Linting 'TableDataSource.swift' (54/125)
Linting 'TableCellItem.swift' (55/125)
Linting 'SimpleTableCellItem.swift' (56/125)
Linting 'TableViewCellRepresentable.swift' (57/125)
Linting 'TableSection.swift' (58/125)
Linting 'ViewRepresentable.swift' (59/125)
Linting 'ControlContainable.swift' (60/125)
Linting 'TableViewContainerCell.swift' (61/125)
Linting 'GridView.swift' (62/125)
Linting 'Actionable.swift' (63/125)
Linting 'DebugMenuConfiguration.swift' (64/125)
Linting 'AppLifecycle.swift' (65/125)
Linting 'LoggingConfiguration.swift' (66/125)
Linting 'StringStyle+GudTay.swift' (67/125)
Linting 'Appearance.swift' (68/125)
Linting 'AppDelegate.swift' (69/125)
Linting 'DebugMenu.swift' (70/125)
Linting 'DebugMenuViewController.swift' (71/125)
/Users/zeisenberg/Projects/gud-tay/GudTay/app/GudTay/Components/Protocols/Actionable.swift:21:9: warning: Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration. (opening_brace)
/Users/zeisenberg/Projects/gud-tay/GudTay/app/GudTay/Components/Protocols/Actionable.swift:27:9: warning: Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration. (opening_brace)
/Users/zeisenberg/Projects/gud-tay/GudTay/app/GudTay/Components/Protocols/Actionable.swift:33:9: warning: Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration. (opening_brace)
Linting 'DebugTextStyleViewController.swift' (72/125)
Linting 'Utilities/BinaryFloatingPoint+Utilities.swift' (73/125)
Linting 'Result.swift' (74/125)
Linting 'CGRect+Utilities.swift' (75/125)
Linting 'Collection+Utilities.swift' (76/125)
Linting 'RandomAccessCollection+Utilities.swift' (77/125)
Linting 'BuildType.swift' (78/125)
Linting 'Asset.swift' (79/125)
Linting 'Services/Utilities/BinaryFloatingPoint+Utilities.swift' (80/125)
/Users/zeisenberg/Projects/gud-tay/GudTay/app/GudTay/Application/Debug/DebugMenu.swift:12:1: warning: Convenience Type Violation: Types used for hosting only static members should be implemented as a caseless enum to avoid instantiation. (convenience_type)
/Users/zeisenberg/Projects/gud-tay/GudTay/app/GudTay/Application/Debug/DebugMenu.swift:37:1: warning: Convenience Type Violation: Types used for hosting only static members should be implemented as a caseless enum to avoid instantiation. (convenience_type)
Linting 'Formatters.swift' (81/125)
Linting 'RequestProtocol.swift' (82/125)
Linting 'APIClient+GudTay.swift' (83/125)
Linting 'APISerialization.swift' (84/125)
Linting 'APIEndpoint+Codable.swift' (85/125)
Linting 'APIError.swift' (86/125)
Linting 'WeatherEndpoint.swift' (87/125)
Linting 'MBTAEndpoint.swift' (88/125)
Linting 'AnyIdentifier.swift' (89/125)
Linting 'FlatCache+Observer.swift' (90/125)
Linting 'FlatCache.swift' (91/125)
Linting 'Identifier.swift' (92/125)
Linting 'JSONAPIResourceType+ImportIncluded.swift' (93/125)
Linting 'JSONAPIPayload.swift' (94/125)
Linting 'JSONEncodable.swift' (95/125)
Linting 'JSONAPI.swift' (96/125)
Linting 'APIClient.swift' (97/125)
Linting 'APIConstants.swift' (98/125)
Linting 'APIEnvironment.swift' (99/125)
Linting 'APIValidation.swift' (100/125)
Linting 'Trip.swift' (101/125)
Linting 'Stop.swift' (102/125)
Linting 'Prediction.swift' (103/125)
Linting 'Route.swift' (104/125)
Linting 'Entity.swift' (105/125)
Linting 'Currently.swift' (106/125)
Linting 'Icon.swift' (107/125)
Linting 'Almanac.swift' (108/125)
Linting 'WeatherBucket.swift' (109/125)
Linting 'Precipitation.swift' (110/125)
Linting 'Temperature.swift' (111/125)
Linting 'Minutely.swift' (112/125)
Linting 'Daily.swift' (113/125)
Linting 'Meteorology.swift' (114/125)
Linting 'WeatherCoordinate.swift' (115/125)
Linting 'WeatherProtocols.swift' (116/125)
Linting 'WeatherForecast.swift' (117/125)
Linting 'NearestStorm.swift' (118/125)
Linting 'Hourly.swift' (119/125)
Linting 'APIEndpoint+Logging.swift' (120/125)
Linting 'APIEndpoint.swift' (121/125)
Linting 'Alamofire+GudTay.swift' (122/125)
Linting 'MBTAService.swift' (123/125)
Linting 'WeatherService.swift' (124/125)
Linting 'LogService.swift' (125/125)
Done linting! Found 5 violations, 0 serious in 125 files.
```

### Environment

* SwiftLint version: 0.38.2
* Installation method used: CocoaPods
* Paste your configuration file:

```yml
disabled_rules:
- cyclomatic_complexity
- empty_parentheses_with_trailing_closure
- file_length
- function_body_length
- function_parameter_count
- large_tuple
- line_length
- nesting
- type_body_length
- identifier_name

opt_in_rules:
  - anyobject_protocol
  - array_init
  - closure_spacing
  - contains_over_filter_count
  - contains_over_filter_is_empty
  - contains_over_first_not_nil
  - contains_over_range_nil_comparison
  - convenience_type
  - empty_count
  - extension_access_modifier
  - fatal_error_message
  - first_where
  - identical_operands
  - implicit_return
  - last_where
  - legacy_random
  - modifier_order
  - multiline_arguments
  - operator_usage_whitespace
  - optional_enum_case_matching
  - prefer_self_type_over_type_of_self
  - redundant_objc_attribute
  - sorted_first_last
  - sorted_imports
  - static_operator
  - toggle_bool
  - unused_control_flow_label
  - unused_import
  - vertical_whitespace

type_name:
  min_length: 2

identifier_name:
    excluded:
        - URL

trailing_comma:
  mandatory_comma: true

statement_position:
  statement_mode: uncuddled_else

custom_rules:
  straight_quotes:
    included: "Localized.swift"
    name: "Straight Quotes"
    regex: "'|\""
    match_kinds:
      - docComment
    message: "Quotes and apostrophes in localized strings should be curly: “” ‘’"
    severity: warning

excluded:
  - Pods
  - GudTay/Resources/Generated
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? **NO**
* Which Xcode version are you using (check `xcodebuild -version`)? **Xcode 11.3 (11C29) on macOS 10.15.2 (19C57)**
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
// This triggers a violation:
class Foo {
    @objc static func foo() {} // can't make it an enum because this needs to be @objc
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -333,7 +333,6 @@ This is the last release to support building with Swift 5.0.x.
   converted to enums.  
   [ZevEisenberg](https://github.com/ZevEisenberg)
   [#3033](https://github.com/realm/SwiftLint/issues/3033)
-* None.
 
 ## 0.38.2: Machine Repair Manual
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -329,6 +329,12 @@ This is the last release to support building with Swift 5.0.x.
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#3100](https://github.com/realm/SwiftLint/issues/3100)
 
+* Fix false positives in `convenience_type` rule for types that cannot be
+  converted to enums.  
+  [ZevEisenberg](https://github.com/ZevEisenberg)
+  [#3033](https://github.com/realm/SwiftLint/issues/3033)
+* None.
+
 ## 0.38.2: Machine Repair Manual
 
 #### Breaking

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -136,6 +136,15 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             return []
         }
 
+        // Has @objc members?
+        let hasObjcMembers = dictionary.substructure.contains { dict in
+            return dict.enclosedSwiftAttributes.contains(.objc)
+        }
+
+        guard !hasObjcMembers else {
+            return []
+        }
+
         return [
             StyleViolation(ruleDescription: Self.description,
                            severity: configuration.severity,

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -35,7 +35,47 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
               public let randomNumber = 2
             }
             """),
-            Example("class DummyClass {}")
+            Example("class DummyClass {}"),
+            Example("""
+            class Foo: NSObject { // class with Obj-C class property
+                class @objc let foo = 1
+            }
+            """),
+            Example("""
+            class Foo: NSObject { // class with Obj-C static property
+                static @objc let foo = 1
+            }
+            """),
+            Example("""
+            class Foo { // non-final class could be inherited
+                class let foo = 1
+            }
+            """),
+            Example("""
+            class Foo { // non-final class with only final members could still be inherited
+                final class let foo = 1
+            }
+            """),
+            Example("""
+            class Foo { // @objc class func can't exist on an enum
+               @objc class func foo() {}
+            }
+            """),
+            Example("""
+            class Foo { // @objc static func can't exist on an enum
+               @objc static func foo() {}
+            }
+            """),
+            Example("""
+            final class Foo { // final class, but @objc class func can't exist on an enum
+               @objc class func foo() {}
+            }
+            """),
+            Example("""
+            final class Foo { // final class, but @objc static func can't exist on an enum
+               @objc static func foo() {}
+            }
+            """)
         ],
         triggeringExamples: [
             Example("""
@@ -52,6 +92,11 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             ↓struct Math {
               public static let pi = 3.14
               @available(*, unavailable) init() {}
+            }
+            """),
+            Example("""
+            final class Foo { // final class can't be inherited
+                class let foo = 1
             }
             """)
         ]

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -84,11 +84,6 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             }
             """),
             Example("""
-            ↓class Math {
-              public static let pi = 3.14
-            }
-            """),
-            Example("""
             ↓struct Math {
               public static let pi = 3.14
               @available(*, unavailable) init() {}

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -100,16 +100,12 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
     public func validate(file: SwiftLintFile, kind: SwiftDeclarationKind,
                          dictionary: SourceKittenDictionary) -> [StyleViolation] {
         guard let offset = dictionary.offset,
-            // MARK: Is class or struct?
             [.class, .struct].contains(kind),
-            // MARK: Inherits from anything?
             dictionary.inheritedTypes.isEmpty,
-            // MARK: Substructure has members? i.e. is the type completely empty?
             !dictionary.substructure.isEmpty else {
                 return []
         }
 
-        // MARK: Contains instance declarations?
         let containsInstanceDeclarations = dictionary.substructure.contains { dict in
             guard let kind = dict.declarationKind else {
                 return false
@@ -131,7 +127,6 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             return []
         }
 
-        // MARK: Has @objc members?
         let hasObjcMembers = dictionary.substructure.contains { dict in
             return dict.enclosedSwiftAttributes.contains(.objc)
         }
@@ -148,12 +143,10 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             ]
         }
 
-        // MARK: Is class?
         guard kind == .class else {
             return makeViolation()
         }
 
-        // MARK: Is final?
         let isFinal = dictionary.swiftAttributes.contains { attributes in
             attributes.attribute == SwiftDeclarationAttributeKind.final.rawValue
         }

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -100,16 +100,16 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
     public func validate(file: SwiftLintFile, kind: SwiftDeclarationKind,
                          dictionary: SourceKittenDictionary) -> [StyleViolation] {
         guard let offset = dictionary.offset,
-            // Is class or struct?
+            // MARK: Is class or struct?
             [.class, .struct].contains(kind),
-            // Inherits from anything?
+            // MARK: Inherits from anything?
             dictionary.inheritedTypes.isEmpty,
-            // Substructure has members? i.e. is the type completely empty?
+            // MARK: Substructure has members? i.e. is the type completely empty?
             !dictionary.substructure.isEmpty else {
                 return []
         }
 
-        // Contains instance declarations?
+        // MARK: Contains instance declarations?
         let containsInstanceDeclarations = dictionary.substructure.contains { dict in
             guard let kind = dict.declarationKind else {
                 return false
@@ -131,7 +131,7 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             return []
         }
 
-        // Has @objc members?
+        // MARK: Has @objc members?
         let hasObjcMembers = dictionary.substructure.contains { dict in
             return dict.enclosedSwiftAttributes.contains(.objc)
         }

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -140,11 +140,26 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             return []
         }
 
-        return [
-            StyleViolation(ruleDescription: Self.description,
-                           severity: configuration.severity,
-                           location: Location(file: file, byteOffset: offset))
-        ]
+        let makeViolation = {
+            return [
+                StyleViolation(ruleDescription: Self.description,
+                               severity: configuration.severity,
+                               location: Location(file: file, byteOffset: offset))
+            ]
+        }
+
+        // MARK: Is class?
+        guard kind == .class else {
+            return makeViolation()
+        }
+
+        // MARK: Is final?
+        let isFinal = dictionary.swiftAttributes.contains { attributes in
+            attributes.attribute == SwiftDeclarationAttributeKind.final.rawValue
+        }
+
+        // Final classes can't be inherited from, so we want to turn it into an enum.
+        return isFinal ? makeViolation() : []
     }
 
     private func isFunctionUnavailable(file: SwiftLintFile, dictionary: SourceKittenDictionary) -> Bool {

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -47,16 +47,6 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             }
             """),
             Example("""
-            class Foo { // non-final class could be inherited
-                class let foo = 1
-            }
-            """),
-            Example("""
-            class Foo { // non-final class with only final members could still be inherited
-                final class let foo = 1
-            }
-            """),
-            Example("""
             class Foo { // @objc class func can't exist on an enum
                @objc class func foo() {}
             }
@@ -92,6 +82,27 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             Example("""
             final ↓class Foo { // final class can't be inherited
                 class let foo = 1
+            }
+            """),
+
+            // Intentional false positives. Non-final classes could be
+            // subclassed, but we figure it is probably rare enough that it is
+            // more important to catch these cases, and manually disable the
+            // rule if needed.
+
+            Example("""
+            ↓class Foo {
+                class let foo = 1
+            }
+            """),
+            Example("""
+            ↓class Foo {
+                final class let foo = 1
+            }
+            """),
+            Example("""
+            ↓class SomeClass {
+                static func foo() {}
             }
             """)
         ]
@@ -135,24 +146,11 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             return []
         }
 
-        let makeViolation = {
-            return [
-                StyleViolation(ruleDescription: Self.description,
-                               severity: configuration.severity,
-                               location: Location(file: file, byteOffset: offset))
-            ]
-        }
-
-        guard kind == .class else {
-            return makeViolation()
-        }
-
-        let isFinal = dictionary.swiftAttributes.contains { attributes in
-            attributes.attribute == SwiftDeclarationAttributeKind.final.rawValue
-        }
-
-        // Final classes can't be inherited from, so we want to turn it into an enum.
-        return isFinal ? makeViolation() : []
+        return [
+            StyleViolation(ruleDescription: Self.description,
+                           severity: configuration.severity,
+                           location: Location(file: file, byteOffset: offset))
+        ]
     }
 
     private func isFunctionUnavailable(file: SwiftLintFile, dictionary: SourceKittenDictionary) -> Bool {

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -90,7 +90,7 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
             }
             """),
             Example("""
-            final class Foo { // final class can't be inherited
+            final ↓class Foo { // final class can't be inherited
                 class let foo = 1
             }
             """)

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ConvenienceTypeRule.swift
@@ -105,12 +105,16 @@ public struct ConvenienceTypeRule: ASTRule, OptInRule, ConfigurationProviderRule
     public func validate(file: SwiftLintFile, kind: SwiftDeclarationKind,
                          dictionary: SourceKittenDictionary) -> [StyleViolation] {
         guard let offset = dictionary.offset,
+            // Is class or struct?
             [.class, .struct].contains(kind),
+            // Inherits from anything?
             dictionary.inheritedTypes.isEmpty,
+            // Substructure has members? i.e. is the type completely empty?
             !dictionary.substructure.isEmpty else {
                 return []
         }
 
+        // Contains instance declarations?
         let containsInstanceDeclarations = dictionary.substructure.contains { dict in
             guard let kind = dict.declarationKind else {
                 return false

